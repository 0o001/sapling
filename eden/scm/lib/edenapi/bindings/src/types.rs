/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

//! These types have a representation that is visible to C / C++, and are either
//! allocated in C / C++ and must be freed accordingly, or do not need to be freed
//! at all (no heap allocations).

use anyhow::Error;
use anyhow::Result;
use edenapi_types::metadata::ContentId as ApiContentId;
use edenapi_types::metadata::Sha1 as ApiSha1;
use edenapi_types::metadata::Sha256 as ApiSha256;
use edenapi_types::FileType as ApiFileType;
use edenapi_types::TreeAttributes as ApiTreeAttributes;
use libc::size_t;
use types::HgId as ApiHgId;
use types::Key as ApiKey;
use types::Parents as ApiParents;
use types::RepoPathBuf;

use crate::ptr_len_to_slice;

#[repr(C)]
pub struct Key {
    path: *const u8,
    path_len: size_t,
    hgid: [u8; 20],
}

// Conversion takes a reference to Key, leaving the caller
// responsible for deallocation of the string.
impl TryFrom<&Key> for ApiKey {
    type Error = Error;
    fn try_from(v: &Key) -> Result<Self> {
        let path = unsafe { ptr_len_to_slice(v.path, v.path_len)? };
        // to_vec copies the slice, so the ApiKey can be
        // deallocated in Rust-land.
        let path = RepoPathBuf::from_utf8(path.to_vec())?;
        Ok(Self {
            path,
            hgid: ApiHgId::from_byte_array(v.hgid),
        })
    }
}

#[repr(C)]
#[derive(Copy, Clone)]
pub enum FileType {
    Regular,
    Executable,
    Symlink,
}

impl From<ApiFileType> for FileType {
    fn from(v: ApiFileType) -> Self {
        match v {
            ApiFileType::Regular => FileType::Regular,
            ApiFileType::Executable => FileType::Executable,
            ApiFileType::Symlink => FileType::Symlink,
        }
    }
}

#[repr(C, u8)]
#[derive(Copy, Clone)]
pub enum Parents {
    None,
    One([u8; 20]),
    Two([u8; 20], [u8; 20]),
}

impl From<ApiParents> for Parents {
    fn from(v: ApiParents) -> Self {
        match v {
            ApiParents::None => Parents::None,
            ApiParents::One(id1) => Parents::One(id1.into_byte_array()),
            ApiParents::Two(id1, id2) => Parents::Two(id1.into_byte_array(), id2.into_byte_array()),
        }
    }
}

#[repr(C)]
#[derive(Copy, Clone)]
pub struct HgId(pub [u8; 20]);

impl From<ApiHgId> for HgId {
    fn from(v: ApiHgId) -> Self {
        HgId(v.into_byte_array())
    }
}

impl From<[u8; 20]> for HgId {
    fn from(v: [u8; 20]) -> Self {
        HgId(v)
    }
}

#[repr(C)]
#[derive(Copy, Clone)]
pub struct ContentId(pub [u8; 32]);

impl From<ApiContentId> for ContentId {
    fn from(v: ApiContentId) -> Self {
        ContentId(v.into())
    }
}

#[repr(C)]
#[derive(Copy, Clone)]
pub struct Sha1(pub [u8; 20]);

impl From<ApiSha1> for Sha1 {
    fn from(v: ApiSha1) -> Self {
        Sha1(v.into())
    }
}

#[repr(C)]
#[derive(Copy, Clone)]
pub struct Sha256(pub [u8; 32]);

impl From<ApiSha256> for Sha256 {
    fn from(v: ApiSha256) -> Self {
        Sha256(v.into())
    }
}

#[repr(C)]
#[derive(Copy, Clone)]
pub struct TreeAttributes {
    manifest_blob: bool,
    parents: bool,
    child_metadata: bool,
}

impl From<TreeAttributes> for ApiTreeAttributes {
    fn from(v: TreeAttributes) -> Self {
        ApiTreeAttributes {
            manifest_blob: v.manifest_blob,
            parents: v.parents,
            child_metadata: v.child_metadata,
        }
    }
}
