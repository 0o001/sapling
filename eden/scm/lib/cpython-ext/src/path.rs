/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

use std::borrow::Borrow;
use std::fmt;
use std::mem;
use std::ops::Deref;
use std::path::Path;
use std::path::PathBuf;

use anyhow::Result;
use cpython::*;
#[cfg(feature = "python2")]
use encoding::local_bytes_to_path;
#[cfg(feature = "python2")]
use encoding::path_to_local_bytes;
use thiserror::Error;
use types::PathComponentBuf;
use types::RepoPath;
use types::RepoPathBuf;

#[cfg(feature = "python2")]
use crate::ResultPyErrExt;

#[derive(Clone, Debug, Eq, PartialEq, PartialOrd, Default, Hash, Ord)]
pub struct PyPathBuf(String);

#[derive(Debug, Eq, PartialEq, PartialOrd, Hash, Ord)]
pub struct PyPath(str);

#[derive(Error, Debug)]
pub enum Error {
    #[error("{0:?} is not a valid UTF-8 path")]
    NonUTF8Path(PathBuf),
}

impl PyPathBuf {
    pub fn as_pypath(&self) -> &PyPath {
        self
    }

    pub fn as_path(&self) -> &Path {
        self.0.as_ref()
    }

    pub fn to_path_buf(&self) -> PathBuf {
        Path::new(&self.0).to_path_buf()
    }

    pub fn to_repo_path_buf(self) -> Result<RepoPathBuf> {
        Ok(RepoPathBuf::from_string(self.0)?)
    }

    pub fn to_repo_path<'a>(&'a self) -> Result<&'a RepoPath> {
        Ok(RepoPath::from_str(&self.0)?)
    }

    pub fn into_utf8_bytes(self) -> Vec<u8> {
        self.0.into()
    }

    pub fn as_utf8_bytes(&self) -> &[u8] {
        self.0.as_bytes()
    }

    pub fn from_string(s: String) -> Self {
        Self(s)
    }

    pub fn from_utf8_bytes(utf8_bytes: Vec<u8>) -> Result<Self> {
        Ok(Self(String::from_utf8(utf8_bytes)?))
    }
}

impl ToPyObject for PyPathBuf {
    #[cfg(feature = "python3")]
    type ObjectType = PyUnicode;
    #[cfg(feature = "python2")]
    type ObjectType = PyBytes;

    #[inline]
    fn to_py_object(&self, py: Python) -> Self::ObjectType {
        #[cfg(feature = "python3")]
        return self.0.to_py_object(py);

        #[cfg(feature = "python2")]
        PyBytes::new(py, &path_to_local_bytes(self.0.as_ref()).unwrap())
    }
}

impl<'source> FromPyObject<'source> for PyPathBuf {
    fn extract(py: Python, obj: &'source PyObject) -> PyResult<Self> {
        #[cfg(feature = "python3")]
        {
            let s = obj.cast_as::<PyUnicode>(py)?.data(py);
            Ok(Self(s.to_string(py)?.into()))
        }

        #[cfg(feature = "python2")]
        {
            let s = obj.cast_as::<PyBytes>(py)?.data(py);
            let path = local_bytes_to_path(s).map_pyerr(py)?;
            Ok(Self(
                path.to_str()
                    .ok_or_else(|| Error::NonUTF8Path(path.to_path_buf()))
                    .map_pyerr(py)?
                    .into(),
            ))
        }
    }
}

impl TryFrom<PathBuf> for PyPathBuf {
    type Error = anyhow::Error;

    fn try_from(path: PathBuf) -> Result<Self> {
        path.as_path().try_into()
    }
}

impl<'a> TryFrom<&'a Path> for PyPathBuf {
    type Error = anyhow::Error;

    fn try_from(path: &'a Path) -> Result<Self> {
        Ok(Self(
            path.to_str()
                .ok_or_else(|| Error::NonUTF8Path(path.to_path_buf()))?
                .into(),
        ))
    }
}

impl From<String> for PyPathBuf {
    fn from(s: String) -> PyPathBuf {
        Self(s)
    }
}

impl<'a> From<&'a RepoPath> for PyPathBuf {
    fn from(repo_path: &'a RepoPath) -> PyPathBuf {
        PyPathBuf(repo_path.as_str().to_owned())
    }
}

impl From<RepoPathBuf> for PyPathBuf {
    fn from(repo_path_buf: RepoPathBuf) -> PyPathBuf {
        PyPathBuf(repo_path_buf.into_string())
    }
}

impl From<PathComponentBuf> for PyPathBuf {
    fn from(path_component_buf: PathComponentBuf) -> PyPathBuf {
        PyPathBuf(path_component_buf.into_string())
    }
}

impl fmt::Display for PyPathBuf {
    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
        fmt::Display::fmt(&*self.0, formatter)
    }
}

impl From<PyPathBuf> for String {
    fn from(path: PyPathBuf) -> String {
        path.0
    }
}

impl Deref for PyPathBuf {
    type Target = PyPath;
    fn deref(&self) -> &Self::Target {
        unsafe { mem::transmute(&*self.0) }
    }
}

impl AsRef<PyPath> for PyPathBuf {
    fn as_ref(&self) -> &PyPath {
        self
    }
}

impl PyPath {
    pub fn from_str(s: &str) -> &Self {
        unsafe { mem::transmute(s) }
    }

    pub fn as_path(&self) -> &Path {
        self.0.as_ref()
    }

    pub fn as_str(&self) -> &str {
        &self.0
    }

    pub fn as_utf8_bytes(&self) -> &[u8] {
        self.0.as_bytes()
    }

    pub fn to_path_buf(&self) -> PathBuf {
        Path::new(&self.0).to_path_buf()
    }

    pub fn to_repo_path(&self) -> Result<&RepoPath> {
        Ok(RepoPath::from_str(&self.0)?)
    }

    pub fn to_repo_path_buf(&self) -> Result<RepoPathBuf> {
        Ok(RepoPathBuf::from_string(self.0.to_string())?)
    }

    pub fn into_utf8_bytes(&self) -> Vec<u8> {
        self.0.into()
    }
}

impl ToOwned for PyPath {
    type Owned = PyPathBuf;

    fn to_owned(&self) -> Self::Owned {
        PyPathBuf(self.0.to_string())
    }
}

impl Borrow<PyPath> for PyPathBuf {
    fn borrow(&self) -> &PyPath {
        self
    }
}

impl RefFromPyObject for PyPath {
    fn with_extracted<F, R>(py: Python, obj: &PyObject, f: F) -> PyResult<R>
    where
        F: FnOnce(&PyPath) -> R,
    {
        #[cfg(feature = "python3")]
        {
            let s = obj.cast_as::<PyUnicode>(py)?.to_string(py)?;
            Ok(f(PyPath::from_str(s.as_ref())))
        }

        #[cfg(feature = "python2")]
        {
            let s = obj.cast_as::<PyBytes>(py)?.data(py);
            let path = local_bytes_to_path(s).map_pyerr(py)?;
            let py_path = PyPath::from_str(
                path.to_str()
                    .ok_or_else(|| Error::NonUTF8Path(path.to_path_buf()))
                    .map_pyerr(py)?,
            );
            Ok(f(py_path))
        }
    }
}

impl AsRef<PyPath> for PyPath {
    fn as_ref(&self) -> &PyPath {
        self
    }
}

impl AsRef<Path> for PyPath {
    fn as_ref(&self) -> &Path {
        self.as_path()
    }
}

impl fmt::Display for PyPath {
    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
        fmt::Display::fmt(&self.0, formatter)
    }
}
