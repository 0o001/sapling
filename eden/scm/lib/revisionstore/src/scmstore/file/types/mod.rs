/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

mod attrs;
mod auxdata;
mod lazy_file;
mod store_file;

pub(crate) use self::lazy_file::LazyFile;
pub use self::{attrs::FileAttributes, auxdata::FileAuxData, store_file::StoreFile};
