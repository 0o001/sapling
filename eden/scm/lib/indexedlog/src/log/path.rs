/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

use crate::errors::IoResultExt;
use crate::lock::ScopedDirLock;
use crate::log::{LogMetadata, META_FILE};
use crate::utils;
use std::path::PathBuf;

/// Abstract Path for [`Log`].
///
/// This defines where a [`Log`] reads and writes data.
#[derive(Clone, Debug)]
pub enum GenericPath {
    /// The [`Log`] is backed by a directory on filesystem.
    Filesystem(PathBuf),

    /// From nothing. Indicates creating from memory.
    Nothing,
}

impl From<&std::path::Path> for GenericPath {
    fn from(path: &std::path::Path) -> Self {
        Self::Filesystem(path.to_path_buf())
    }
}

impl From<&str> for GenericPath {
    fn from(path: &str) -> Self {
        Self::Filesystem(std::path::Path::new(path).to_path_buf())
    }
}

impl From<PathBuf> for GenericPath {
    fn from(path: PathBuf) -> Self {
        Self::Filesystem(path)
    }
}

impl From<&PathBuf> for GenericPath {
    fn from(path: &PathBuf) -> Self {
        Self::Filesystem(path.clone())
    }
}

impl From<()> for GenericPath {
    fn from(_path: ()) -> Self {
        Self::Nothing
    }
}

impl GenericPath {
    /// Return the main filesystem path.
    pub fn as_opt_path(&self) -> Option<&std::path::Path> {
        match self {
            GenericPath::Filesystem(path) => Some(&path),
            GenericPath::Nothing => None,
        }
    }

    pub(crate) fn mkdir(&self) -> crate::Result<()> {
        if let Some(dir) = self.as_opt_path() {
            utils::mkdir_p(dir)
        } else {
            Ok(())
        }
    }

    pub(crate) fn lock(&self) -> crate::Result<ScopedDirLock> {
        if let Some(dir) = self.as_opt_path() {
            Ok(ScopedDirLock::new(&dir)?)
        } else {
            Err(crate::Error::programming(
                "read_meta() does not support GenericPath::Nothing",
            ))
        }
    }

    pub(crate) fn read_meta(&self) -> crate::Result<LogMetadata> {
        match self {
            GenericPath::Filesystem(dir) => {
                let meta_path = dir.join(META_FILE);
                LogMetadata::read_file(&meta_path).context(&meta_path, "when reading LogMetadata")
            }
            GenericPath::Nothing => Err(crate::Error::programming(
                "read_meta() does not support GenericPath::Nothing",
            )),
        }
    }

    pub(crate) fn write_meta(&self, meta: &LogMetadata, fsync: bool) -> crate::Result<()> {
        match self {
            GenericPath::Filesystem(dir) => {
                let meta_path = dir.join(META_FILE);
                meta.write_file(&meta_path, fsync)?;
                Ok(())
            }
            GenericPath::Nothing => Err(crate::Error::programming(
                "write_meta() does not support GenericPath::Nothing",
            )),
        }
    }
}
