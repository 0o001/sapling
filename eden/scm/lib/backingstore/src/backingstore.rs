/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

use crate::remotestore::FakeRemoteStore;
use crate::treecontentstore::TreeContentStore;
use crate::utils::key_from_path_node_slice;
use anyhow::Result;
use edenapi::{Builder as EdenApiBuilder, EdenApi};
use manifest::{List, Manifest};
use manifest_tree::TreeManifest;
use revisionstore::{
    ContentStore, ContentStoreBuilder, EdenApiFileStore, EdenApiTreeStore, HgIdDataStore,
    LegacyStore, LocalStore, RemoteDataStore, StoreKey, StoreResult,
};
use std::path::Path;
use std::sync::Arc;
use types::{Key, Node, RepoPathBuf};

pub struct BackingStore {
    blobstore: ContentStore,
    treestore: Arc<TreeContentStore>,
}

impl BackingStore {
    pub fn new<P: AsRef<Path>>(repository: P, use_edenapi: bool) -> Result<Self> {
        let hg = repository.as_ref().join(".hg");
        let config = configparser::hg::load::<String, String>(Some(&hg), None)?;

        let store_path = hg.join("store");
        #[allow(unused_mut)]
        let mut blobstore = ContentStoreBuilder::new(&config).local_path(&store_path);
        let treestore = ContentStoreBuilder::new(&config)
            .local_path(&store_path)
            .suffix(Path::new("manifests"));

        // Memcache takes 30s to initialize on debug builds slowing down tests significantly, let's
        // not even try to initialize it then.
        #[cfg(not(debug_assertions))]
        {
            use log::warn;
            use progress::null::NullProgressFactory;
            use revisionstore::MemcacheStore;
            match MemcacheStore::new(&config, NullProgressFactory::arc()) {
                Ok(memcache) => {
                    // XXX: Add the memcachestore for the treestore.
                    blobstore = blobstore.memcachestore(Arc::new(memcache));
                }
                Err(e) => warn!("couldn't initialize Memcache: {}", e),
            }
        }

        let (blobstore, treestore) = match config.get_opt::<String>("remotefilelog", "reponame")? {
            Some(repo) if use_edenapi => {
                let edenapi = EdenApiBuilder::from_config(&config)?.build()?;
                let edenapi: Arc<dyn EdenApi> = Arc::new(edenapi);
                let fileremotestore = EdenApiFileStore::new(repo.clone(), edenapi.clone(), None);
                let treeremotestore = EdenApiTreeStore::new(repo, edenapi, None);
                (
                    blobstore.remotestore(fileremotestore).build()?,
                    treestore.remotestore(treeremotestore).build()?,
                )
            }
            _ => (
                blobstore.remotestore(Arc::new(FakeRemoteStore)).build()?,
                treestore.remotestore(Arc::new(FakeRemoteStore)).build()?,
            ),
        };

        Ok(Self {
            blobstore,
            treestore: Arc::new(TreeContentStore::new(treestore)),
        })
    }

    fn get_blob_impl(&self, key: Key) -> Result<Option<Vec<u8>>> {
        // Return None for LFS blobs
        // TODO: LFS support
        if let Ok(StoreResult::Found(metadata)) =
            self.blobstore.get_meta(StoreKey::hgid(key.clone()))
        {
            if metadata.is_lfs() {
                return Ok(None);
            }
        }

        Ok(self
            .blobstore
            .get_file_content(&key)?
            .map(|blob| blob.as_ref().to_vec()))
    }

    /// Reads file from blobstores. When `local_only` is true, this function will only read blobs
    /// from on disk stores.
    pub fn get_blob(&self, path: &[u8], node: &[u8], local_only: bool) -> Result<Option<Vec<u8>>> {
        let key = key_from_path_node_slice(path, node)?;

        // check if the blob present on disk
        if local_only && !self.blobstore.contains(&StoreKey::from(&key))? {
            return Ok(None);
        }

        self.get_blob_impl(key)
    }

    /// Fetch file contents in batch. Whenever a blob is fetched, the supplied `resolve` function is
    /// called with the file content or an error message, and the index of the blob in the request
    /// array. When `local_only` is enabled, this function will only check local disk for the file
    /// content.
    pub fn get_blob_batch<F>(&self, keys: Vec<Result<Key>>, local_only: bool, resolve: F)
    where
        F: Fn(usize, Result<Option<Vec<u8>>>) -> (),
    {
        // logic:
        // 1. convert all path & nodes into `StoreKey`
        // 2. try to resolve blobs that are already local
        // 3. fetch anything that is not local and refetch
        let requests = keys
            .into_iter()
            .enumerate()
            .filter_map(|(index, key)| match key {
                Ok(key) => Some((index, key)),
                Err(e) => {
                    // return early when the key is invalid
                    resolve(index, Err(e));
                    None
                }
            });

        let mut missing = Vec::new();
        let mut missing_requests = Vec::new();

        for (index, key) in requests {
            let store_key = StoreKey::from(&key);
            // Assuming a blob do not exist if `.contains` call fails
            if self.blobstore.contains(&store_key).unwrap_or(false) {
                resolve(index, self.get_blob_impl(key))
            } else if !local_only {
                missing.push(store_key);
                missing_requests.push((index, key));
            }
        }

        // If this is a local only read, nothing else we can do.
        if local_only {
            return;
        }

        let _ = self.blobstore.prefetch(&missing);
        for (index, key) in missing_requests {
            resolve(index, self.get_blob_impl(key))
        }
    }

    fn get_tree_impl(&self, key: Key) -> Result<List> {
        let manifest = TreeManifest::durable(self.treestore.clone(), key.hgid);
        manifest.list(&key.path)
    }

    pub fn get_tree(&self, node: &[u8], local_only: bool) -> Result<Option<List>> {
        let node = Node::from_slice(node)?;
        let path = RepoPathBuf::new();
        let key = Key::new(path, node);

        // check if the blob is present on disk
        if local_only
            && !self
                .treestore
                .as_content_store()
                .contains(&StoreKey::from(&key))?
        {
            return Ok(None);
        }

        Ok(Some(self.get_tree_impl(key)?))
    }

    /// Fetch tree contents in batch. Whenever a tree is fetched, the supplied `resolve` function is
    /// called with the tree content or an error message, and the index of the tree in the request
    /// array. When `local_only` is enabled, this function will only check local disk for the file
    /// content.
    pub fn get_tree_batch<F>(&self, keys: Vec<Result<Key>>, local_only: bool, resolve: F)
    where
        F: Fn(usize, Result<Option<List>>) -> (),
    {
        // logic:
        // 1. convert all path & nodes into `StoreKey`
        // 2. try to resolve blobs that are already local
        // 3. fetch anything that is not local and refetch
        let requests = keys
            .into_iter()
            .enumerate()
            .filter_map(|(index, key)| match key {
                Ok(key) => Some((index, key)),
                Err(e) => {
                    // return early when the key is invalid
                    resolve(index, Err(e));
                    None
                }
            });

        let mut missing = Vec::new();
        let mut missing_requests = Vec::new();

        let contentstore = self.treestore.as_content_store();
        for (index, key) in requests {
            let store_key = StoreKey::from(&key);
            // Assuming a blob do not exist if `.contains` call fails
            if contentstore.contains(&store_key).unwrap_or(false) {
                resolve(index, Some(self.get_tree_impl(key)).transpose())
            } else if !local_only {
                missing.push(store_key);
                missing_requests.push((index, key));
            }
        }

        // If this is a local only read, nothing else we can do.
        if local_only {
            return;
        }

        let _ = contentstore.prefetch(&missing);
        for (index, key) in missing_requests {
            resolve(index, Some(self.get_tree_impl(key)).transpose())
        }
    }

    /// Forces backing store to rescan pack files or local indexes
    pub fn refresh(&self) {
        self.blobstore.refresh().ok();
        self.treestore.as_content_store().refresh().ok();
    }
}
