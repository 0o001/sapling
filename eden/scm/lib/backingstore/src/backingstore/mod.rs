/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

mod contentstores;
mod scmstores;

use std::path::Path;

use anyhow::Result;

use manifest::List;
use types::Key;

use crate::backingstore::{contentstores::BackingContentStores, scmstores::BackingScmStores};

pub enum BackingStore {
    Old(BackingContentStores),
    New(BackingScmStores),
}

use BackingStore::*;

impl BackingStore {
    pub fn new<P: AsRef<Path>>(repository: P, use_edenapi: bool) -> Result<Self> {
        let hg = repository.as_ref().join(".hg");
        let config = configparser::hg::load::<String, String>(Some(&hg), None)?;

        Ok(if config.get_or_default("scmstore", "backingstore")? {
            New(BackingScmStores::new(&config, &hg, use_edenapi)?)
        } else {
            Old(BackingContentStores::new(&config, &hg, use_edenapi)?)
        })
    }

    /// Reads file from blobstores. When `local_only` is true, this function will only read blobs
    /// from on disk stores.
    pub fn get_blob(&self, path: &[u8], node: &[u8], local_only: bool) -> Result<Option<Vec<u8>>> {
        match self {
            Old(stores) => stores.get_blob(path, node, local_only),
            New(stores) => stores.get_blob(path, node, local_only),
        }
    }

    /// Fetch file contents in batch. Whenever a blob is fetched, the supplied `resolve` function is
    /// called with the file content or an error message, and the index of the blob in the request
    /// array. When `local_only` is enabled, this function will only check local disk for the file
    /// content.
    pub fn get_blob_batch<F>(&self, keys: Vec<Result<Key>>, local_only: bool, resolve: F)
    where
        F: Fn(usize, Result<Option<Vec<u8>>>) -> (),
    {
        match self {
            Old(stores) => stores.get_blob_batch(keys, local_only, resolve),
            New(stores) => stores.get_blob_batch(keys, local_only, resolve),
        }
    }

    pub fn get_tree(&self, node: &[u8], local_only: bool) -> Result<Option<List>> {
        match self {
            Old(stores) => stores.get_tree(node, local_only),
            New(stores) => stores.get_tree(node, local_only),
        }
    }

    /// Fetch tree contents in batch. Whenever a tree is fetched, the supplied `resolve` function is
    /// called with the tree content or an error message, and the index of the tree in the request
    /// array. When `local_only` is enabled, this function will only check local disk for the file
    /// content.
    pub fn get_tree_batch<F>(&self, keys: Vec<Result<Key>>, local_only: bool, resolve: F)
    where
        F: Fn(usize, Result<Option<List>>) -> (),
    {
        match self {
            Old(stores) => stores.get_tree_batch(keys, local_only, resolve),
            New(stores) => stores.get_tree_batch(keys, local_only, resolve),
        }
    }

    /// Forces backing store to write its pending data to disk and to read the latest version from
    /// the disk.
    pub fn refresh(&self) {
        match self {
            Old(stores) => stores.refresh(),
            New(stores) => stores.refresh(),
        }
    }
}

impl Drop for BackingStore {
    fn drop(&mut self) {
        // Make sure that all the data that was fetched is written to the hgcache.
        self.refresh();
    }
}
