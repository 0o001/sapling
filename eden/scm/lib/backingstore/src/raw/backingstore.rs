/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

//! Provides the c-bindings for `crate::backingstore`.

use std::slice;
use std::str;

use anyhow::ensure;
use anyhow::Error;
use anyhow::Result;
use libc::c_char;
use libc::c_void;
use libc::size_t;
use manifest::List;
use revisionstore::scmstore::FileAuxData as ScmStoreFileAuxData;
use types::Key;

use crate::backingstore::BackingStore;
use crate::raw::CBytes;
use crate::raw::CFallible;
use crate::raw::FileAuxData;
use crate::raw::Request;
use crate::raw::Tree;

fn stringpiece_to_slice<'a, T, U>(ptr: *const T, length: size_t) -> Result<&'a [U]> {
    ensure!(!ptr.is_null(), "string ptr is null");
    Ok(unsafe { slice::from_raw_parts(ptr as *const U, length) })
}

fn backingstore_new(
    repository: *const c_char,
    repository_len: size_t,
    use_edenapi: bool,
    aux_data: bool,
) -> Result<*mut BackingStore> {
    super::init::backingstore_global_init();

    let repository = stringpiece_to_slice(repository, repository_len)?;
    let repo = str::from_utf8(repository)?;
    let store = Box::new(BackingStore::new(repo, use_edenapi, aux_data)?);

    Ok(Box::into_raw(store))
}

#[no_mangle]
pub extern "C" fn rust_backingstore_new(
    repository: *const c_char,
    repository_len: size_t,
    use_edenapi: bool,
    aux_data: bool,
) -> CFallible<BackingStore> {
    backingstore_new(repository, repository_len, use_edenapi, aux_data).into()
}

#[no_mangle]
pub extern "C" fn rust_backingstore_free(store: *mut BackingStore) {
    assert!(!store.is_null());
    let store = unsafe { Box::from_raw(store) };
    drop(store);
}

fn backingstore_get_blob(
    store: *mut BackingStore,
    name: *const u8,
    name_len: usize,
    node: *const u8,
    node_len: usize,
    local: bool,
) -> Result<*mut CBytes> {
    assert!(!store.is_null());
    let store = unsafe { &*store };
    let path = stringpiece_to_slice(name, name_len)?;
    let node = stringpiece_to_slice(node, node_len)?;

    store
        .get_blob(path, node, local)
        .and_then(|opt| opt.ok_or_else(|| Error::msg("no blob found")))
        .map(CBytes::from_vec)
        .map(|result| Box::into_raw(Box::new(result)))
}

#[no_mangle]
pub extern "C" fn rust_backingstore_get_blob(
    store: *mut BackingStore,
    name: *const u8,
    name_len: usize,
    node: *const u8,
    node_len: usize,
    local: bool,
) -> CFallible<CBytes> {
    backingstore_get_blob(store, name, name_len, node, node_len, local).into()
}

#[no_mangle]
pub extern "C" fn rust_backingstore_get_blob_batch(
    store: *mut BackingStore,
    requests: *const Request,
    size: usize,
    local: bool,
    data: *mut c_void,
    resolve: unsafe extern "C" fn(*mut c_void, usize, CFallible<CBytes>),
) {
    assert!(!store.is_null());
    let store = unsafe { &*store };
    let requests: &[Request] = unsafe { slice::from_raw_parts(requests, size) };
    let keys: Vec<Result<Key>> = requests.iter().map(|req| req.try_into_key()).collect();

    store.get_blob_batch(keys, local, |idx, result| {
        let result = result
            .and_then(|opt| opt.ok_or_else(|| Error::msg("no blob found")))
            .map(CBytes::from_vec)
            .map(|result| Box::into_raw(Box::new(result)));
        unsafe {
            resolve(data, idx, result.into())
        };
    });
}

fn backingstore_get_tree(
    store: *mut BackingStore,
    node: *const u8,
    node_len: usize,
    local: bool,
) -> Result<*mut Tree> {
    assert!(!store.is_null());
    let store = unsafe { &*store };
    let node = stringpiece_to_slice(node, node_len)?;

    store
        .get_tree(node, local)
        .and_then(|opt| opt.ok_or_else(|| Error::msg("no tree found")))
        .and_then(|list| list.try_into())
        .map(|result| Box::into_raw(Box::new(result)))
}

#[no_mangle]
pub extern "C" fn rust_backingstore_get_tree(
    store: *mut BackingStore,
    node: *const u8,
    node_len: usize,
    local: bool,
) -> CFallible<Tree> {
    backingstore_get_tree(store, node, node_len, local).into()
}

#[no_mangle]
pub extern "C" fn rust_backingstore_get_tree_batch(
    store: *mut BackingStore,
    requests: *const Request,
    size: usize,
    local: bool,
    data: *mut c_void,
    resolve: unsafe extern "C" fn(*mut c_void, usize, CFallible<Tree>),
) {
    assert!(!store.is_null());
    let store = unsafe { &*store };
    let requests: &[Request] = unsafe { slice::from_raw_parts(requests, size) };
    let keys: Vec<Result<Key>> = requests.iter().map(|req| req.try_into_key()).collect();

    store.get_tree_batch(keys, local, |idx, result| {
        let result: Result<List> =
            result.and_then(|opt| opt.ok_or_else(|| Error::msg("no tree found")));
        let result: Result<Tree> = result.and_then(|list| list.try_into());
        let result: Result<*mut Tree> = result.map(|result| Box::into_raw(Box::new(result)));
        unsafe {
            resolve(data, idx, result.into())
        };
    });
}

fn backingstore_get_file_aux(
    store: *mut BackingStore,
    node: *const u8,
    node_len: usize,
    local: bool,
) -> Result<*mut FileAuxData> {
    assert!(!store.is_null());
    let store = unsafe { &*store };
    let node = stringpiece_to_slice(node, node_len)?;

    store
        .get_file_aux(node, local)
        .and_then(|opt| opt.ok_or_else(|| Error::msg("no file aux data found")))
        .map(|aux| aux.into())
        .map(|result| Box::into_raw(Box::new(result)))
}

#[no_mangle]
pub extern "C" fn rust_backingstore_get_file_aux(
    store: *mut BackingStore,
    node: *const u8,
    node_len: usize,
    local: bool,
) -> CFallible<FileAuxData> {
    backingstore_get_file_aux(store, node, node_len, local).into()
}

#[no_mangle]
pub extern "C" fn rust_backingstore_get_file_aux_batch(
    store: *mut BackingStore,
    requests: *const Request,
    size: usize,
    local: bool,
    data: *mut c_void,
    resolve: unsafe extern "C" fn(*mut c_void, usize, CFallible<FileAuxData>),
) {
    assert!(!store.is_null());
    let store = unsafe { &*store };
    let requests: &[Request] = unsafe { slice::from_raw_parts(requests, size) };
    let keys: Vec<Result<Key>> = requests.iter().map(|req| req.try_into_key()).collect();

    store.get_file_aux_batch(keys, local, |idx, result| {
        let result: Result<ScmStoreFileAuxData> =
            result.and_then(|opt| opt.ok_or_else(|| Error::msg("no file aux data found")));
        let result: Result<FileAuxData> = result.map(|aux| aux.into());
        let result: Result<*mut FileAuxData> = result.map(|result| Box::into_raw(Box::new(result)));
        unsafe {
            resolve(data, idx, result.into())
        };
    });
}

#[no_mangle]
pub extern "C" fn rust_tree_free(tree: *mut Tree) {
    assert!(!tree.is_null());
    let tree = unsafe { Box::from_raw(tree) };
    drop(tree);
}

#[no_mangle]
pub extern "C" fn rust_file_aux_free(aux: *mut FileAuxData) {
    assert!(!aux.is_null());
    let aux = unsafe { Box::from_raw(aux) };
    drop(aux);
}

#[no_mangle]
pub extern "C" fn rust_backingstore_flush(store: *mut BackingStore) {
    assert!(!store.is_null());
    let store = unsafe { &*store };

    store.flush();
}
