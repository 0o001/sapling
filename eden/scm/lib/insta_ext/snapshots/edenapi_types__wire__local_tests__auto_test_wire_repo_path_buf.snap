---
source: mod.rs
expression: "WireRepoPathBuf::arbitrary(&mut g)"

---
"erlvio/goxk/vl/qvjafti/vh/xha"
